codegen/type_resolution.py

```diff
@@ -1038,12 +1040,27 @@ def select_function_with_least_implicit_conversion_cost(
             [(fn.format_for_output_to_user(), fn.loc) for _, fn in best_functions],
         )
 
+    def add_function_to_codegen(
+        self, declaration: FunctionDeclaration, signature: FunctionSignature
+    ) -> None:
+        for other in self._already_codegened[declaration.uuid]:
+            if (
+                do_specializations_match(other.specialization, signature.specialization)
+                and other.arguments == signature.arguments
+                and other.return_type == signature.return_type
+            ):
+                return  # Already codegened
+
+        self._already_codegened[declaration.uuid].append(signature)
+        self._remaining_to_codegen.append((declaration, signature))
```

Comment:
Perhaps `_already_codegened` should be a set then

Reply:
Yeah, this would be a valid optimization. Atm it would be kinda annoying to freeze the signature and add the hashing/ equality comparisons so its linear complexity for now :upside_down_face:.

Reply:
Let's do it at some point though. Just changing `do_specializations_match()` to list equality is a 5.5% performance boost, if we do even fewer comparisons by caching hashes etc then it'll add up very quickly